File: Selection.py
# Sure enough this is a module

# define a class
class Choice:

    def __init__(self):
        self.i = 0

    def groupSelection(self, candidates, size):
        self.words = candidates
        self.r = size
        self.complete_group = []
        self.i = 0
        
        self.recursiveFillUp([])

        return self.complete_group

    # pick elements recursively
    def recursiveFillUp(self, temp):
        picked_elements = [[] for k in range(len(self.words))]
        j = self.i
        while j < len(self.words):
            picked_elements[j].extend(temp)
            picked_elements[j].append(self.words[j])
            # recoil factor
            if self.i >= len(self.words):
                self.i = j
                
            # satisfied yet?
            if len(picked_elements[j]) == self.r:
                self.complete_group.append(picked_elements[j])
            elif len(picked_elements[j]) < self.r:
                self.recursiveFillUp(picked_elements[j])
            
            j += 1
        
        j -= 1
        if picked_elements[j] != None and len(picked_elements[j]) == self.r:
            self.i += 1 # keep recoil factor straightened out
        

File: Selection.cpp
#include "stdafx.h"
#include "Selection.h"


Selection::Selection()
{
}

vector<vector<string>> Selection::groupSelection(vector<string> candidates, unsigned short dimension) {
	words = candidates;
	r = dimension;
	complete_group = {};
	i = 0;
	recursiveFillUp({});

	return complete_group;
}

// pick elements recursively
void Selection::recursiveFillUp(vector<string> temp) {
	vector<string> * picked_elements;
	picked_elements = new vector<string>[words.size()];
	unsigned j = i;
	while (j < words.size()) {
		picked_elements[j] = temp;
		picked_elements[j].push_back(words[j]);
		// recoil factor
		if (i >= words.size()) {
			i = j;
		}
		// satisfied yet?
		if (picked_elements[j].size() == r) {
			complete_group.push_back(picked_elements[j]);
		}
		else if (picked_elements[j].size() < r) {
			recursiveFillUp(picked_elements[j]);
		}
		j++;
	}
	if (!picked_elements[--j].empty() && picked_elements[j].size() == r) {
		i++; // keep recoil factor straightened out
	}
	delete[] picked_elements;
}

Selection::~Selection()
{
}
